File: internal/models/pipeline.go
package models

import (
	"fmt"
	"strconv"
	"time"

	"github.com/clintjedwards/gofer/internal/storage"
	sdk "github.com/clintjedwards/gofer/sdk/go/config"
	"github.com/rs/zerolog/log"
)

// A collection of logically grouped tasks. A task is a unit of work wrapped in a docker container.
// Pipeline is a secondary level unit being contained within namespaces and containing runs.
type Pipeline struct {
	Metadata PipelineMetadata `json:"metadata" doc:"Macro-level details on the targeted pipeline"`
	Config   PipelineConfig   `json:"config" doc:"User controlled data on the targeted pipeline"`
}

type PipelineState string

const (
	PipelineStateUnknown  PipelineState = "UNKNOWN"
	PipelineStateActive   PipelineState = "ACTIVE"
	PipelineStateDisabled PipelineState = "DISABLED"
)

// Details about the pipeline itself, not including the configuration that the user can change.
// All these values are changed by the system or never changed at all. This sits in contrast to
// the config which the user can change freely.
type PipelineMetadata struct {
	NamespaceID string        `json:"namespace_id" example:"default" doc:"Unique identifier of the target namespace"`
	PipelineID  string        `json:"pipeline_id" example:"simple_pipeline" doc:"Unique identifier for the target pipeline; user supplied"`
	Created     uint64        `json:"created" example:"1712433802634" doc:"Time of pipeline creation in epoch milliseconds"`
	Modified    uint64        `json:"modified" example:"1712433802634" doc:"Time pipeline was updated to a new version in epoch milliseconds."`
	State       PipelineState `json:"state" example:"ACTIVE" doc:"The current running state of the pipeline. This is used to determine if the pipeline should continue to process runs or not and properly convey that to the user."`
}

func NewPipelineMetadata(namespace, id string) *PipelineMetadata {
	newPipelineMetadata := &PipelineMetadata{
		NamespaceID: namespace,
		PipelineID:  id,
		Created:     uint64(time.Now().UnixMilli()),
		Modified:    uint64(time.Now().UnixMilli()),
		State:       PipelineStateActive,
	}

	return newPipelineMetadata
}

func (p *PipelineMetadata) ToStorage() *storage.PipelineMetadata {
	return &storage.PipelineMetadata{
		Namespace: p.NamespaceID,
		ID:        p.PipelineID,
		Created:   fmt.Sprint(p.Created),
		Modified:  fmt.Sprint(p.Modified),
		State:     string(p.State),
	}
}

func (p *PipelineMetadata) FromStorage(sp *storage.PipelineMetadata) {
	created, err := strconv.ParseUint(sp.Created, 10, 64)
	if err != nil {
		log.Fatal().Err(err).Msg("error in translating from storage")
	}

	modified, err := strconv.ParseUint(sp.Modified, 10, 64)
	if err != nil {
		log.Fatal().Err(err).Msg("error in translating from storage")
	}

	p.NamespaceID = sp.Namespace
	p.PipelineID = sp.ID
	p.Created = created
	p.Modified = modified
	p.State = PipelineState(sp.State)
}

type PipelineConfigState string

const (
	PipelineConfigStateUnknown    PipelineConfigState = "UNKNOWN"
	PipelineConfigStateUnreleased PipelineConfigState = "UNRELEASED" // Has never been deployed.
	PipelineConfigStateLive       PipelineConfigState = "LIVE"       // Currently deployed.
	PipelineConfigStateDeprecated PipelineConfigState = "DEPRECATED" // Has previously been deployed and is now defunct.
)

// A representation of the user's configuration settings for a particular pipeline.
type PipelineConfig struct {
	NamespaceID string              `json:"namespace_id" example:"default" doc:"Unique identifier of the target namespace"`
	PipelineID  string              `json:"pipeline_id" example:"simple_pipeline" doc:"Unique identifier for the target pipeline; user supplied"`
	Version     int64               `json:"version" example:"42" doc:"The current version of this pipeline configuration"`
	Parallelism int64               `json:"parallelism" example:"5" doc:"The amount of runs allowed to happen at any given time"`
	Name        string              `json:"name" example:"Simple Pipeline" doc:"Human readable name for pipeline"`
	Description string              `json:"description" example:"Some description here" doc:"Description of pipeline's purpose and other details"`
	Tasks       map[string]Task     `json:"tasks" doc:"Tasks associated with this pipeline config"`
	State       PipelineConfigState `json:"state" example:"LIVE" doc:"The deployment state of the config. This is used to determine if the pipeline should continue to process runs or not and properly convey that to the user"`
	Registered  uint64              `json:"registered" example:"1712433802634" doc:"Time in epoch milliseconds when this pipeline config was registered"`
	Deprecated  uint64              `json:"deprecated" example:"1712433802634" doc:"The time in epoch milliseconds when the pipeline config was marked deprecated. This helps us figure out which version is actually most recent and which are defunct"`
}

func NewPipelineConfig(namespace, pipeline string, version int64, config *sdk.UserPipelineConfig) *PipelineConfig {
	tasks := map[string]Task{}

	for _, taskRaw := range config.Tasks {
		ct := Task{}
		ct.FromSDKUserPipelineTaskConfig(taskRaw)
		tasks[ct.ID] = ct
	}

	return &PipelineConfig{
		NamespaceID: namespace,
		PipelineID:  pipeline,
		Version:     version,
		Parallelism: config.Parallelism,
		Name:        config.Name,
		Description: config.Description,
		Tasks:       tasks,
		State:       PipelineConfigStateUnreleased,
		Registered:  uint64(time.Now().UnixMilli()),
		Deprecated:  0,
	}
}

func (pc *PipelineConfig) ToStorage() (*storage.PipelineConfig, []*storage.PipelineTask) {
	pipelineConfig := &storage.PipelineConfig{
		Namespace:   pc.NamespaceID,
		Pipeline:    pc.PipelineID,
		Version:     pc.Version,
		Parallelism: pc.Parallelism,
		Name:        pc.Name,
		Description: pc.Description,
		Registered:  fmt.Sprint(pc.Registered),
		Deprecated:  fmt.Sprint(pc.Deprecated),
		State:       string(pc.State),
	}

	tasks := []*storage.PipelineTask{}

	for _, task := range pc.Tasks {
		tasks = append(tasks, task.ToStorage(pc.NamespaceID, pc.PipelineID, pc.Version))
	}

	return pipelineConfig, tasks
}

func (pc *PipelineConfig) FromStorage(spc *storage.PipelineConfig, spct *[]storage.PipelineTask,
) {
	tasks := map[string]Task{}

	for _, task := range *spct {
		var ct Task
		ct.FromStorage(&task)
		tasks[task.ID] = ct
	}

	registered, err := strconv.ParseUint(spc.Registered, 10, 64)
	if err != nil {
		log.Fatal().Err(err).Msg("error in translating from storage")
	}

	deprecated, err := strconv.ParseUint(spc.Deprecated, 10, 64)
	if err != nil {
		log.Fatal().Err(err).Msg("error in translating from storage")
	}

	pc.NamespaceID = spc.Namespace
	pc.PipelineID = spc.Pipeline
	pc.Version = spc.Version
	pc.Parallelism = spc.Parallelism
	pc.Name = spc.Name
	pc.Description = spc.Description
	pc.Tasks = tasks
	pc.State = PipelineConfigState(spc.State)
	pc.Registered = registered
	pc.Deprecated = deprecated
}

File: gofer_types/src/api/pipeline.rs
use super::pipeline_config;
use crate::{epoch_milli, storage};
use anyhow::{Context, Result};
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use std::str::FromStr;
use strum::{Display, EnumString};

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
pub struct PipelinePathArgsRoot {
    /// The unique identifier for the target namespace.
    pub namespace_id: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
pub struct PipelinePathArgs {
    /// The unique identifier for the target namespace.
    pub namespace_id: String,

    /// The unique identifier for the target pipeline.
    pub pipeline_id: String,
}

#[derive(
    Debug, Clone, Display, Default, PartialEq, EnumString, Eq, Serialize, Deserialize, JsonSchema,
)]
#[serde(rename_all = "snake_case")]
#[strum(serialize_all = "snake_case")]
#[strum(ascii_case_insensitive)]
pub enum PipelineState {
    #[default]
    Unknown,

    Active,

    Disabled,
}

/// Details about the pipeline itself, not including the configuration that the user can change.
/// All these values are changed by the system or never changed at all. This sits in contrast to
/// the config which the user can change freely.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, JsonSchema)]
pub struct Metadata {
    /// Unique identifier of the target namespace.
    pub namespace_id: String,

    /// Unique identifier of the target pipeline.
    pub pipeline_id: String,

    /// Time of pipeline creation in epoch milliseconds.
    pub created: u64,

    /// Time pipeline was updated to a new version in epoch milliseconds.
    pub modified: u64,

    /// The current running state of the pipeline. This is used to determine if the pipeline should run or not.
    pub state: PipelineState,
}

impl Metadata {
    pub fn new(namespace_id: &str, pipeline_id: &str) -> Self {
        Metadata {
            namespace_id: namespace_id.into(),
            pipeline_id: pipeline_id.into(),
            created: epoch_milli(),
            modified: 0,
            state: PipelineState::Active,
        }
    }
}

impl TryFrom<storage::pipeline_metadata::PipelineMetadata> for Metadata {
    type Error = anyhow::Error;

    fn try_from(value: storage::pipeline_metadata::PipelineMetadata) -> Result<Self> {
        let created = value.created.parse::<u64>().with_context(|| {
            format!(
                "Could not parse field 'created' from storage value '{}'",
                value.created
            )
        })?;

        let modified = value.modified.parse::<u64>().with_context(|| {
            format!(
                "Could not parse field 'modified' from storage value '{}'",
                value.modified
            )
        })?;

        let state = PipelineState::from_str(&value.state).with_context(|| {
            format!(
                "Could not parse field 'token type' from storage value '{}'",
                value.state
            )
        })?;

        Ok(Metadata {
            namespace_id: value.namespace_id,
            pipeline_id: value.pipeline_id,
            created,
            modified,
            state,
        })
    }
}

impl From<Metadata> for storage::pipeline_metadata::PipelineMetadata {
    fn from(value: Metadata) -> Self {
        Self {
            namespace_id: value.namespace_id,
            pipeline_id: value.pipeline_id,
            created: value.created.to_string(),
            modified: value.modified.to_string(),
            state: value.state.to_string(),
        }
    }
}

/// A collection of logically grouped tasks. A task is a unit of work wrapped in a docker container.
/// Pipeline is a secondary level unit being contained within namespaces and containing runs.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, JsonSchema)]
pub struct Pipeline {
    /// Macro level details on the targeted pipeline.
    pub metadata: Metadata,

    /// User controlled data for the targeted pipeline.
    pub config: pipeline_config::Config,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
pub struct ListPipelinesResponse {
    /// A list of all pipelines metadata.
    pub pipelines: Vec<Metadata>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
pub struct GetPipelineResponse {
    /// The metadata for the pipeline.
    pub pipeline: Metadata,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
pub struct UpdatePipelineRequest {
    pub state: Option<PipelineState>,
}
